### PR TITLE
TextBox updates

### LINKED ISSUE #9433: TextBox programmatic Undo/Redo
**Is your feature request related to a problem? Please describe.**
TextBox currently has no way to programatically call Undo or Redo or determine if undo/redo is available. While probably not widely useful, it is needed for something like TextCommandBarFlyout (or menus) where they're triggered by menu items. Currently I'm just raising key events to trigger it, but that isn't the most optimal solution and I also don't have any way of detecting if Undo/Redo has entries.

Proposed API, based on the UWP TextBox API which has this functionality
https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.textbox?view=winrt-22621

```diff
public class TextBox
{
// Read-only bool properties to determine if the Undo/Redo stacks have entries available
+    DirectProperty<TextBox, bool> CanUndoProperty;
+    DirectProperty<TextBox, bool> CanRedoProperty;

+    public bool CanUndo { get; }
+    public bool CanRedo { get; }

// Methods to call Undo or Redo programatically
+    public void Undo()
+    public void Redo()
}
```

### PR DESCRIPTION
**Pull request by amwx**
## What does the pull request do?
Primarily, this PR adds support for programmatic undo/redo and adds `CanUndo` and `CanRedo` properties. Also added a couple unit tests for undo/redo (there were none before), and added a bunch of missing xml docs for public members on TextBox. 

`CanUndo`/`CanRedo` was implemented by adding two methods to `UndoRedoHelper.IUndoRedoHost` that `UndoRedoHelper` calls when the stacks are modified (`OnUndoStackChanged` and `OnRedoStackChanged`). TextBox implements these to update the properties, which are `DirectProperty` to support binding in menus (for example).

## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Fixed issues
Fixes #9433

### CODE REVIEW

**Review by maxkatz6 — APPROVED** [review]

**Comment by amwx** [comment]
`Keys_Allow_Undo` test is failing when run with `Tab` and `Enter`. The test is failing with `Tab` and `Enter` because `CanUndo` is reporting false.

This line in `UndoRedoHelper` is failing because current is equal to _currentNode, so the snapshot is never taken.
```C#
var current = _host.UndoRedoState;
if (_currentNode == null || !_currentNode.Value!.Equals(current))
```

I'll be honest, I find the undo/redo logic as implemented really confusing and hard to follow - and it seems to have some hacks in places, so I really don't know what the correct solution for this is. This also seems to be more a hidden existing bug, but I'm not really sure how this passed before because the only modification I made was to check `CanUndo`.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0026485-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by amwx** [comment]
I took another look at this, and I think I found a solution. In the setter for `Text`, the snapshot was being taken after the text property was changed - meaning the "initial" state wasn't set correctly and you essentially had to make 2 text changes to get the undo stack to unlock - at least that's based on my understanding, I could be wrong and this may have other unintended side effects. The test passes now, but I don't know why the integration tests are - they don't seem related to any changes I made

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0026496-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Review by maxkatz6 — APPROVED** [review]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0026540-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]
